### PR TITLE
bpo-39994: Fix pprint handling of dict subclasses that override __repr__

### PR DESCRIPTION
**Pull request by iritkatriel**
https://bugs.python.org/issue39994

Co-authored-by: Palak Kumar Jha

### CODE REVIEW

File: Lib/test/test_pprint.py [hunk 1]
```diff
@@ -187,6 +208,21 @@ def test_same_as_repr(self):
...
+        for simple in (list_custom_repr([]), list_custom_repr([1,2,3]), list_custom_repr(range(N)),
+                       tuple_custom_repr([]), tuple_custom_repr([1,2,3]), tuple_custom_repr(range(N)),
+                       set_custom_repr([]), set_custom_repr([1,2,3]), set_custom_repr(range(N)),
+                       frozenset_custom_repr([]), frozenset_custom_repr([1,2,3]),  frozenset_custom_repr(range(N)),
+                       dict_custom_repr(), dict_custom_repr({5: 6}), dict_custom_repr(zip(range(N),range(N))),
+                      ):
+            native = repr(simple)
+            self.assertEqual(pprint.pformat(simple), '*' * len(native))
+            self.assertEqual(pprint.pformat(simple, width=1, indent=0)
+                             .replace('\n', ' '), '*' * len(native))
```
Comment (serhiy-storchaka) [inline Lib/test/test_pprint.py]: Is `.replace('\n', ' ')` really needed?

File: Lib/test/test_pprint.py [hunk 2]
```diff
@@ -187,6 +208,21 @@ def test_same_as_repr(self):
                              .replace('\n', ' '), native)
             self.assertEqual(pprint.saferepr(simple), native)
 
+    def test_repr_override_called(self):
+        N = 1000
+        # Ensure that a __repr__ override is called for subclasses of containers.
+        for simple in (list_custom_repr([]), list_custom_repr([1,2,3]), list_custom_repr(range(N)),
```
Comment (serhiy-storchaka) [inline Lib/test/test_pprint.py]: These lines seem too long. PEP 8 sets limit of line length for code to 79 characters.

File: Misc/NEWS.d/next/Library/2020-08-15-18-17-21.bpo-39994.dOgPOh.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fixed pprint's handling of dict subclasses that override __repr__
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2020-08-15-18-17-21.bpo-39994.dOgPOh.rst]: Add a sentence ending period.

File: Lib/test/test_pprint.py (hunk 1 above)
Comment (iritkatriel) [reply Lib/test/test_pprint.py]: Indeed not needed, I'll remove.

File: Lib/test/test_pprint.py (hunk 2 above)
Comment (iritkatriel) [reply Lib/test/test_pprint.py]: Thanks, will fix.

File: Misc/NEWS.d/next/Library/2020-08-15-18-17-21.bpo-39994.dOgPOh.rst (hunk 3 above)
Comment (iritkatriel) [reply Misc/NEWS.d/next/Library/2020-08-15-18-17-21.bpo-39994.dOgPOh.rst]: Will do.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**